### ISSUE TITLE
Regression due to #5558

### ISSUE THREAD

**Issue by David-16** [micronaut-projects/micronaut-core#6089]
### Expected Behavior

The attached wiremock test will timeout with `io.micronaut.http.client.exceptions.ReadTimeoutException: Read Timeout`

### Actual Behaviour

The attached wiremock test should pass

### Steps To Reproduce

```
import com.github.tomakehurst.wiremock.WireMockServer;
import com.github.tomakehurst.wiremock.client.WireMock;
import com.github.tomakehurst.wiremock.core.WireMockConfiguration;
import io.micronaut.buffer.netty.NettyByteBufferFactory;
import io.micronaut.http.HttpRequest;
import io.micronaut.http.HttpResponse;
import io.micronaut.http.MediaType;
import io.micronaut.http.MutableHttpRequest;
import io.micronaut.http.client.RxStreamingHttpClient;
import io.reactivex.Flowable;
import org.junit.jupiter.api.Test;

import java.net.URL;
import java.nio.charset.StandardCharsets;

import static com.github.tomakehurst.wiremock.client.WireMock.*;


public class WiremockTests {
    @Test
    void testWiremock() throws Exception {
        final WireMockServer wireMockServer = new WireMockServer(WireMockConfiguration.wireMockConfig().dynamicPort().dynamicHttpsPort());
        wireMockServer.start();

        WireMock.configureFor("http", "0.0.0.0", wireMockServer.port());
        WireMock.configureFor("https", "0.0.0.0", wireMockServer.httpsPort());

        String mockUrl = "/";
        stubFor(put(urlEqualTo(mockUrl))
            .inScenario("my-scenario")
            .willSetStateTo("created")
            .willReturn(aResponse().withStatus(200).withBody("contents")));

        byte[] bytes = "fdsfdsfadf".getBytes(StandardCharsets.UTF_8);
        Flowable<?> bufs = Flowable.just(NettyByteBufferFactory.DEFAULT.wrap(bytes));

        MutableHttpRequest<?> request = HttpRequest.PUT(mockUrl, bufs)
            .contentType(MediaType.APPLICATION_OCTET_STREAM_TYPE)
            .accept(MediaType.ALL_TYPE)
            .contentLength(bytes.length);

        HttpResponse<Void> response = RxStreamingHttpClient
            .create(new URL("https://0.0.0.0:" + wireMockServer.httpsPort()))
            .exchange(request, Void.class)
            .singleOrError()
            .blockingGet();

        System.out.println(response.getStatus());
    }
}
```

Debugging the issue, it was found that wiremock servlet is waiting to read the body.  After more debugging, issue was narrowed down to a change in Micronaut 2.5.5 (test passes) and 2.5.6 (test fails)

Specifically due to https://github.com/micronaut-projects/micronaut-core/pull/5558 'Delay emission of response headers until at least 1 item is emitted in a streaming response'

Even more specifically, due to the `ctx.writeAndFlush` which was removed: https://github.com/micronaut-projects/micronaut-core/commit/0f6cec56d635326911077fb221df40f0c4d9fbe6#diff-65d4d2d0808b6551855bc6a632d8fea706e6158e9904f80b129c0351e7d6e229L394

Though `ctx.writeAndFlush` was added to the reactive callback and even though it is getting called, it does not seem to be working the same way and so the wiremock servlet timesout trying to read the body

I do not think there is anything wiremock specific to the problem, because adding the following to `HttpStreamsHandler` makes it all work again even in the latest Micronaut 2.5.12:
```
ctx.writeAndFlush(message);
messageWritten.compareAndSet(false, true);
```

I have tried to figure out where things are going wrong, but I cannot seem to find a meaningful delta even though it is clearly there

### Environment Information

CentOS Linux
JDK 11

### Version

2.5.6-2.5.12

**Comment by jameskleeh** [comment]
Are you seeing this issue in any real world scenario?

**Comment by David-16** [comment]
I did try similar unit test cases against httpbin.org and those seemed to pass which makes me wonder if it is something on the wiremock side

However, the issue arose while trying to upgrade from Micronaut 2.5.5 to 2.5.12.  Since unit tests failed (due to aforementioned wiremock test), I cannot confirm if it happens in real life without building a separate release that is potentially broken, setting up a separate test bed, yadda, yadda, and unfortunately that takes a lot of time and care

Still, let me see if I can turn out a separate out-of-band service that can isolate the issue yet still talks to a production-like instance

**Comment by jameskleeh** [comment]
@David-16 Sounds good. I just don't think this is worth a time investment if it only fails with wiremock. We can't revert to the previous behavior and the current logic works in all of our tests that use a real server

**Comment by jameskleeh** [comment]
@David-16 Can you try with 3.0.2?

**Comment by aleksandrsvoboda** [comment]
This happened to me recently as well. We are using a @Filter annotation from micronaut to proxy requests. After upgrading from micronaut 2.5.5 to anything above POST and PUT requests in wiremock tests return with `io.micronaut.http.client.exceptions.ReadTimeoutException: Read Timeout`. When running the actual service the same POST and PUT calls just never finish executing. Probably a bad solution, but a quick and easy one is to force the dependency for the following two packages:
```
io.micronaut:micronaut-http-netty:2.5.5
io.micronaut:micronaut-http-server-netty:2.5.5
```
In gradle:
```
configurations.all {
  resolutionStrategy {
    force 'io.micronaut:micronaut-http-netty:2.5.5'
    force 'io.micronaut:micronaut-http-server-netty:2.5.5'
  }
}
```

**Comment by jameskleeh** [comment]
@aleksandrsvoboda If you can reproduce in a sample application please upload it here

**Comment by aleksandrsvoboda** [comment]
@jameskleeh will do, thanks for a quick reply.

**Comment by aleksandrsvoboda** [comment]
https://github.com/aleksandrsvoboda/sample-project
Here is a sample project. Clone, run and call 
```
curl --request POST --url http://localhost:8080/proxy/posts --header 'Content-Type: application/json' --data '{"id": 123}'
```

The request above will never finish. However if you call either of the requests below they will succeed. Adding a body to the POST request breaks this for some reason. Also, if you change the micronaut version to 2.5.5 in `gradle.properties` all of these requests succeed.

```
curl --request POST --url http://localhost:8080/proxy/posts --header 'Content-Type: application/json'
```
```
curl --request GET --url http://localhost:8080/proxy/posts --header 'Content-Type: application/json'
```

**Comment by David-16** [comment]
Apologies.  This has mostly fallen off my radar as we are going through production pains.  We do have plans for upgrading to 3.x but probably not for a couple months

@aleksandrsvoboda Did you try the `HttpStreamsHandler` workaround mentioned in the original bug report?  I did not see anything specific to streaming in your project, but I am no expert so am curious if it works for you as well

**Comment by aleksandrsvoboda** [comment]
@David-16 I have not tried this. I'm pretty inexperienced at this type of things and I did not quite understand your solution. Did you mean I should add the two lines below to `HttpStreamsHandler` around line `410`?
```
ctx.writeAndFlush(message);
messageWritten.compareAndSet(false, true);
```
Also, would I have to fork the repo since `HttpStreamsHandler` is read only?

**Comment by David-16** [comment]
Yes, putting this line back helped me (line 394 in the old file / line 411 in the new file): https://github.com/micronaut-projects/micronaut-core/commit/0f6cec56d635326911077fb221df40f0c4d9fbe6#diff-65d4d2d0808b6551855bc6a632d8fea706e6158e9904f80b129c0351e7d6e229L394

In the copy I had, `messageWritten` was defined outside the anonymous class but looks like this might have changed since.  If so, then maybe you can ignore the `messageWritten` line or you may have to move it to be defined outside the anonymous class

To overwrite, I simply copied the class using the same package name in my library and made the changes.  The class is then compiled and included in the jar generated by my project.  As long as it shows up first in the classpath (e.g., it is included before micronaut jars), then the JVM will load your copy first.  This works process generally works unless the jars are sealed/signed or when using modules (probably)

**Comment by jameskleeh** [comment]
This should be resolved with 3.0.2
